File: src/upros_tutorial/src/ros_server.cpp
#include "ros/ros.h"
#include "upros_message/MyServiceMsgRequest.h"
#include "upros_message/MyServiceMsgResponse.h"
#include "upros_message/MyServiceMsg.h"

bool myServiceCallback(upros_message::MyServiceMsgRequest &req, upros_message::MyServiceMsgResponse &res) {
    // 处理服务请求
    res.output = req.input * 2;
    ROS_INFO("Request: input = %d, output = %d", req.input, res.output);
    return true;
}

int main(int argc, char **argv) {
    // 初始化ROS节点
    ros::init(argc, argv, "my_service_node");
    // 创建节点句柄
    ros::NodeHandle nh;
    // 创建服务
    ros::ServiceServer service = nh.advertiseService("my_service", myServiceCallback);
    
    ROS_INFO("Ready to receive service requests.");

    // 循环处理ROS回调函数
    ros::spin();
    return 0;
}

File: src/upros_tutorial/scripts/ros_server.py
#!/usr/bin/env python3

import rospy
from upros_message.srv import MyServiceMsg

def handle_my_service(req):
    rospy.loginfo("Service request: %d", req.input)
    return req.input * 2  # 返回输出结果

rospy.init_node("my_service_server")
service = rospy.Service("my_service", MyServiceMsg, handle_my_service)
rospy.loginfo("Ready to handle requests")
rospy.spin()
